umap/distances.py

```diff
@@ -182,6 +182,43 @@ def minkowski_grad(x, y, p=2):
     return result ** (1.0 / p), grad
 
 
+@numba.njit()
+def poincare(u, v):
+    """Poincare distance.
+
+    ..math::
+        \delta (u, v) = 2 \frac{ \lVert  u - v \rVert ^2 }{ ( 1 - \lVert  u \rVert ^2 ) ( 1 - \lVert  v \rVert ^2 ) }
+        D(x, y) = \operatorname{arcosh} (1+\delta (u,v))
```

Comment:
Out of curiosity, does this generalise to u and v not 2-dimensional? That is, does this validly define a notional "Poincare ball"?

Reply:
yep, as i understand it, it does generalize to >2 dimensions

Reply:
Sweet! It looked like it should, but that is definitely a nice bonus.